connector/src/main/scala/com/microsoft/kusto/spark/utils/ContainerProvider.scala

```diff
@@ -158,80 +170,110 @@ class ContainerProvider(
     roundRobinIdx = 0
     storage(roundRobinIdx)
   }
-}
 
-object ContainerProvider {
-  private val className = this.getClass.getSimpleName
-  private val sasKeyCacheMap = new ConcurrentHashMap[String,ContainerAndSas]()
+  private val sasKeyCacheMap = new ConcurrentHashMap[String, ContainerAndSas]()
   def refreshUserSas(
-      ingestionStorageParams: Array[IngestionStorageParameters],
-      isCacheExpired:Boolean,
-      cacheExpirySeconds: Long,
-      listPermissions:Boolean=false): ContainerAndSas = {
+                      ingestionStorageParams: Array[IngestionStorageParameters],
+                      isCacheExpired: Boolean,
+                      cacheExpirySeconds: Long,
+                      listPermissions: Boolean = false): (Boolean, ContainerAndSas) = {
 
     val ingestionStorageParameter =
       IngestionStorageParameters.getRandomIngestionStorage(ingestionStorageParams)
 
     val key = ingestionStorageParameter.toString
 
     // If the cache has not expired and the key is already in the cache, return the cached value
-    KDSU.logInfo("ContainerProvider",s" Checking cache for Key: $key")
-    if(!isCacheExpired && sasKeyCacheMap.contains(key)) {
-      ContainerAndSas(
-        s"${ingestionStorageParameter.storageUrl}/${ingestionStorageParameter.containerName}",
-        s"?${sasKeyCacheMap.get(key)}")
+    KDSU.logInfo("ContainerProvider", s" Checking cache for Key: $key")
```

Comment:
its just a little too many calls to normalizeSasKey  
if its in the cache or if we generated it - it should be already normalized
if its from the user we can do it once in the KDSU when parsing the user options